fdbserver/Resolver.actor.cpp

```diff
@@ -18,36 +18,75 @@
  * limitations under the License.
  */
 
-#include "flow/ActorCollection.h"
 #include "fdbclient/NativeAPI.actor.h"
-#include "fdbserver/ResolverInterface.h"
-#include "fdbserver/MasterInterface.h"
-#include "fdbserver/WorkerInterface.actor.h"
-#include "fdbserver/WaitFailure.h"
+#include "fdbclient/SystemData.h"
+#include "fdbserver/ConflictSet.h"
 #include "fdbserver/Knobs.h"
-#include "fdbserver/ServerDBInfo.h"
+#include "fdbserver/MasterInterface.h"
 #include "fdbserver/Orderer.actor.h"
-#include "fdbserver/ConflictSet.h"
+#include "fdbserver/PartMerger.h"
+#include "fdbserver/ResolverInterface.h"
+#include "fdbserver/ServerDBInfo.h"
 #include "fdbserver/StorageMetrics.h"
-#include "fdbclient/SystemData.h"
+#include "fdbserver/TimedKVCache.h"
+#include "fdbserver/WaitFailure.h"
+#include "fdbserver/WorkerInterface.actor.h"
+#include "flow/ActorCollection.h"
+
 #include "flow/actorcompiler.h"  // This must be the last #include.
 
+#include "debug.h"
+
 namespace {
+
+class ResolveTransactionBatchRequestMerger : public PartMerger<ResolveTransactionBatchRequest> {
+protected:
+	virtual void mergeFirstPart(const value_t& incomingRequest) override {
+		ASSERT(incomingRequest.transactions.size() == 1);
+		merged = incomingRequest;
+	}
+
+	virtual void merge(const value_t& incomingRequest) override {
+		ASSERT(incomingRequest.transactions.size() == 1);
+
+		auto& arena = merged.arena;
+		auto& current = merged.transactions[0];
+		auto& incoming = incomingRequest.transactions[0];
+
+		ASSERT(current.read_snapshot == incoming.read_snapshot);
+		ASSERT(current.report_conflicting_keys == incoming.report_conflicting_keys);
+
+		// COUT << "Received " << incomingRequest.splitTransaction.get().id << " part " << incomingRequest.splitTransaction.get().partIndex << std::endl;
+
+		merged.arena.dependsOn(incomingRequest.arena);
+
+		current.read_conflict_ranges.append(arena, incoming.read_conflict_ranges.begin(),
+		                                    incoming.read_conflict_ranges.size());
+		current.write_conflict_ranges.append(arena, incoming.write_conflict_ranges.begin(),
+		                                     incoming.write_conflict_ranges.size());
+		// Mutations are not useful in the resolve context
```

Comment:
This assumption is wrong -- there are mutations for txnStateTransaction that are piggybacked in the ResolveTransactionBatchRequest, which is meant to be sent back to other proxies. So here we need to merge these mutations as well.

The merging process needs to preserve the order, i.e., part number. This ordering requirement is a bit tricky, since the current merge doesn't preserve the order.